contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,251 @@
+    // reset every epoch
+    access(account) var nodeHasSubmitted: {String: Bool}
+
+    /// Array of unique final submissions from nodes
+    /// if a final submission is sent that matches one that already has been submitted
+    /// this array will not change at all
+    access(account) var uniqueFinalSubmissions: [[String]]
+
+    /// Tracks how many submissions have been sent
+    /// for each unique final submission
+    access(account) var uniqueFinalSubmissionCount: {Int: UInt64}
+
+    // ================================================================================
+    // CONTRACT CONSTANTS
+    // ================================================================================
+
+    // Canonical paths for admin and participant resources
+    pub let AdminStoragePath: StoragePath
+    pub let ParticipantStoragePath: StoragePath
+
+    /// Struct to represent a single whiteboard message
+    pub struct Message {
+
+        pub let nodeID: String
+
+        pub let phase: UInt8
+
+        pub let content: String
+
+        init(nodeID: String, phase: UInt8, content: String) {
+            self.nodeID = nodeID
+            self.phase = phase
+            self.content = content
+        }
+    }
+
+    // The Participant resource is generated for each consensus node when they register.
+    // Each resource instance is good for all future potential epochs, but will
+    // only be valid if the node operator has been confirmed as a consensus node for the next epoch.
+    pub resource Participant {
+
+        pub let nodeID: String
+
+        // Submits a whiteboard message to the contract
+        pub fun sendMessage(phase: UInt8, _ content: String) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send whiteboard message if not registered for the current epoch"
+                phase > FlowDKG.currentPhase && phase < UInt8(4): "Phase submission is invalid"
+            }
+
+            // create the message struct
+            let message = Message(nodeID: self.nodeID, phase: phase, content: content)
+
+            // add the message to the message record for the phase
+            FlowDKG.phaseMessages[phase]!.append(message)
+
+            emit BroadcastMessage(phase: phase, nodeID: self.nodeID, content: content)
+
+        }
+
+        /// Sends the final key vector submission. 
+        /// Can only be called during the final phase and by consensus nodes that are registered.
+        pub fun sendFinalSubmission(_ submission: [String]) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send final submission if not registered for the current epoch"
+                !FlowDKG.nodeHasSubmittedFinal(self.nodeID)!: "Cannot submit a final submission twice"
+                FlowDKG.currentPhase == DKGPhase.finalSubmission: "Can only send final submission in the final DKG phase"
+                // submission should be a certain length?
```

Comment:
It should be `n+1` where `n` is the number of participants in the current DKG instance -- I think `n` would be `FlowDKG.nodeHasSubmitted.keys.length`